cli/cmd/agent_aws-install_ec2ssm.go

```diff
@@ -0,0 +1,249 @@
+	)
+	agentInstallAWSSSMCmd.Flags().StringVar(
+		&agentCmdState.InstallBYORole,
+		"iam_role_name",
+		"",
+		"IAM role name (not ARN) with SSM policy, if not provided then an ephemeral role will be created",
+	)
+}
+
+func installAWSSSM(_ *cobra.Command, _ []string) error {
+	token := agentCmdState.InstallAgentToken
+	if token == "" {
+		if cli.InteractiveMode() {
+			// user didn't provide an agent token
+			cli.Log.Debugw("agent token not provided, asking user to select one now")
+			var err error
+			token, err = selectAgentAccessToken()
+			if err != nil {
+				return err
+			}
+		} else {
+			return errors.New("user did not provide or interactively select an agent token")
+		}
+	}
+
+	runners, err := awsDescribeInstances()
+	if err != nil {
+		return err
+	}
+
+	cfg, err := GetConfig()
+	if err != nil {
+		return err
+	}
+	role, instanceProfile, err := SetupSSMAccess(cfg, agentCmdState.InstallBYORole, token)
+	defer func() {
+		err := TeardownSSMAccess(cfg, role, instanceProfile, agentCmdState.InstallBYORole) // clean up after ourselves
+		cli.Log.Warnw("got an error while tearing down IAM infra", "error", err)
+	}()
+	if err != nil {
+		return err
+	}
+
+	wg := new(sync.WaitGroup)
+	wp := workerpool.New(agentCmdState.InstallMaxParallelism)
+	for _, runner := range runners {
+		wg.Add(1)
+
+		// In order to use `wp.Submit()`, the input func() must not take any arguments.
+		// Copy the runner info to dedicated variable in the goroutine to prevent race overwrite
+		runnerCopyWg := new(sync.WaitGroup)
+		runnerCopyWg.Add(1)
+
+		wp.Submit(func() {
+			defer wg.Done()
+
+			threadRunner := *runner
+			runnerCopyWg.Done()
+
+			cli.Log.Debugw("runner info: ",
+				"user", threadRunner.Runner.User,
+				"region", threadRunner.Region,
+				"az", threadRunner.AvailabilityZone,
+				"instance_id", threadRunner.InstanceID,
+				"hostname", threadRunner.Runner.Hostname,
+			)
+
+			// Attach an instance profile with our new role to the runner
+			err = threadRunner.AssociateInstanceProfileWithRunner(cfg, instanceProfile)
+			if err != nil {
+				cli.Log.Debugw("failed to attach instance profile to runner",
+					"error", err,
+					"instance ID", threadRunner.InstanceID,
+					"role", role,
+					"instance profile", instanceProfile,
+				)
+				return
+			} else {
+				cli.OutputHuman(fmt.Sprintf("successfully associated with instance ID %s\n", threadRunner.InstanceID))
+			}
+
+			// Establish SSM Command connection to the runner
+
+			// Check if agent is already installed on the host, skip if yes
+			// Sleep for up to 5min to wait for instance profile to associate with instance
+			var ssmError error
+			var commandOutput ssm.GetCommandInvocationOutput
+			for i := 0; i < 5; i++ {
+				cli.Log.Debugw("waiting for instance profile to associate with instance, sleeping 1min",
+					"iteration number (time slept in minutes)", i,
+					"instance ID", threadRunner.InstanceID,
+				)
+				time.Sleep(1 * time.Minute)
```

Comment:
it seems like this should be at the end of the loop to only execute on failure, (and I guess check for `i < 6` to keep it for 5 minutes)

Reply:
This is a good point. Here are the tradeoffs as I see them:

Pros: 
* Only sleep on failure (if customer has instance profile already associated, this saves them a minute of execution)

Cons:
* The last minute of sleep isn't useful, since we've already tried the command and failed, so we're waiting a minute just to return an error. 

I think the solution that has the best of both worlds is to wrap the sleep with a conditional that executes the sleep if it's not the last iteration in the loop. Which seems a little messy to me, but sleeps aren't exactly pristine code in the first place. 

I'm open to other suggestions too!

Reply:
Done with 84b48008fe1f12807fad27b1546d1499e38dafb5

Reply:
Yeah I think the largest issue is actually that if you have 1000 hosts and a workgroup of size 10 then you will sleep for about 100 minutes, even though you don't need to sleep for more than 1 minute -- the sleep should only need to happen the first time through.

Your fix looks good to me.